src/data/SeriesData.ts

```diff
@@ -0,0 +1,1338 @@
+            const dimIdx = this._getStoreDimIndex(dim);
+            innerRange[dimIdx] = range[dim];
+            dimIndices.push(dimIdx);
+        });
+
+        if (__DEV__) {
+            validateDimensions(this, dimIndices);
+        }
+
+        this._store = this._store.clone();
+        this._store.selectRange(innerRange);
+        return this;
+    }
+
+    /**
+     * Data mapping to a plain array
+     */
+    mapArray<Ctx, Cb extends MapArrayCb0<Ctx>>(cb: Cb, ctx?: Ctx, ctxCompat?: Ctx): ReturnType<Cb>[];
+    /* eslint-disable */
+    mapArray<Ctx, Cb extends MapArrayCb1<Ctx>>(dims: DimensionLoose, cb: Cb, ctx?: Ctx, ctxCompat?: Ctx): ReturnType<Cb>[];
+    mapArray<Ctx, Cb extends MapArrayCb1<Ctx>>(dims: [DimensionLoose], cb: Cb, ctx?: Ctx, ctxCompat?: Ctx): ReturnType<Cb>[];
+    mapArray<Ctx, Cb extends MapArrayCb2<Ctx>>(dims: [DimensionLoose, DimensionLoose], cb: Cb, ctx?: Ctx, ctxCompat?: Ctx): ReturnType<Cb>[];
+    mapArray<Ctx, Cb extends MapArrayCb<Ctx>>(dims: ItrParamDims, cb: Cb, ctx?: Ctx, ctxCompat?: Ctx): ReturnType<Cb>[];
+    /* eslint-enable */
+    mapArray<Ctx>(
+        dims: ItrParamDims | MapArrayCb<Ctx>,
+        cb: MapArrayCb<Ctx> | Ctx,
+        ctx?: Ctx
+    ): unknown[] {
+        'use strict';
+
+        if (typeof dims === 'function') {
+            ctx = cb as Ctx;
+            cb = dims;
+            dims = [];
+        }
+
+        // ctxCompat just for compat echarts3
+        ctx = (ctx || this) as Ctx;
+
+        const result: unknown[] = [];
+        this.each(dims, function () {
+            result.push(cb && (cb as MapArrayCb<Ctx>).apply(this, arguments));
+        }, ctx);
+        return result;
+    }
+
+    /**
+     * Data mapping to a new List with given dimensions
+     */
+    map<Ctx>(dims: DimensionLoose, cb: MapCb1<Ctx>, ctx?: Ctx, ctxCompat?: Ctx): SeriesData<HostModel>;
+    map<Ctx>(dims: [DimensionLoose], cb: MapCb1<Ctx>, ctx?: Ctx, ctxCompat?: Ctx): SeriesData<HostModel>;
+    /* eslint-disable-next-line */
+    map<Ctx>(dims: [DimensionLoose, DimensionLoose], cb: MapCb2<Ctx>, ctx?: Ctx, ctxCompat?: Ctx): SeriesData<HostModel>;
+    map<Ctx>(
+        dims: ItrParamDims,
+        cb: MapCb<Ctx>,
+        ctx?: Ctx,
+        ctxCompat?: Ctx
+    ): SeriesData {
+        'use strict';
+
+        // ctxCompat just for compat echarts3
+        const fCtx = (ctx || ctxCompat || this) as CtxOrList<Ctx>;
+
+        const dimIndices = map(
+            normalizeDimensions(dims), this._getStoreDimIndex, this
+        );
+
+        if (__DEV__) {
+            validateDimensions(this, dimIndices);
+        }
+
+        const list = cloneListForMapAndSample(this);
+        list._store = this._store.map(dimIndices, (fCtx
+            ? zrUtil.bind(cb as any, fCtx as any)
+            : cb) as any
+        );
+        return list;
+    }
+
+    /**
+     * Large data down sampling on given dimension
+     * @param sampleIndex Sample index for name and id
+     */
+    downSample(
```

Comment:
### Just a programming pattern thinking

Since we have shared data in `DataStorage` instead of `SeriesData/List`, 
should we change the interface like `SeriesData#downSamle`/`SeriesData#lttbDownSample`/`SeriesData#map` from immutable to mutable (only provide immutable interface in `DataStorage`),
And substitute `SeriesData#reset` for `SeriesData#cloneShallow`.
And remove `Series#setData`.
Finally a `series` instance will not change the reference of `seriesData` instance,
and the file `src/data/helper/linkSeriesData.ts`, which is a little weird, can be removed.
And the concept might be more simple: only `DataStorage` can be shared and immutable, rather than both `DataStorage` and `SeriesData` need to be  immutable.

Reply:
Good point. It will be much simpler if we can reuse the same Data instance.

But for `DataStorage`. I'm thinking it should be the opposite. That all methods should be immutable. So if in one series do filter / downSample. It won't affect other Series which share the same DataStorage.

Reply:
Yes I agree, `DataStorage` will be totally immutable. (The only exception is `appendData` and `appendValue`, which are restricted to only be called in `Series#getRawData`.